httpserver/mcp_advanced_integration_test.go

```diff
@@ -0,0 +1,521 @@
+
+	if accessToken == "" {
+		t.Fatal("Failed to get access token with dynamically registered client")
+	}
+
+	t.Logf("Successfully obtained token with dynamic client")
+
+	// Test MCP access with the token
+	t.Log("Testing MCP access with dynamic client token...")
+	testMCPInitialize(t, client, baseURL, accessToken)
+
+	t.Log("Dynamic client registration test passed!")
+}
+
+// TestMCPWithSSO tests MCP access via SSO (mock SSO server with authorization code flow)
+func TestMCPWithSSO(t *testing.T) {
+	// Skip if condor_master is not available
+	if _, err := exec.LookPath("condor_master"); err != nil {
+		t.Skip("condor_master not found in PATH, skipping integration test")
+	}
+
+	// Setup main MCP server
+	tempDir, mcpServer, mcpBaseURL, signingKeyPath := setupTestServer(t, 18083)
+	defer os.RemoveAll(tempDir)
+	defer shutdownTestServer(t, mcpServer)
+
+	// Setup mock SSO server
+	ssoPort := 18084
+	ssoServer := setupMockSSOServer(t, ssoPort, tempDir)
+	ssoBaseURL := fmt.Sprintf("http://127.0.0.1:%d", ssoPort)
+	defer shutdownMockSSOServer(t, ssoServer)
+
+	client := &http.Client{Timeout: 30 * time.Second}
+	testUser := "ssouser"
+	testPassword := "ssopassword"
+
+	// Create OAuth2 client on MCP server pointing to SSO
+	t.Log("Creating OAuth2 client for SSO flow...")
+	storage := mcpServer.GetOAuth2Provider().GetStorage()
+
+	clientID := "sso-test-client"
+	clientSecret := "sso-test-secret"
+
+	oauth2Client := &fosite.DefaultClient{
+		ID:            clientID,
+		Secret:        []byte(clientSecret),
+		RedirectURIs:  []string{mcpBaseURL + "/callback"},
+		GrantTypes:    []string{"authorization_code", "refresh_token"},
+		ResponseTypes: []string{"code"},
+		Scopes:        []string{"openid", "mcp:read", "mcp:write"},
+		Public:        false,
+	}
+
+	if err := storage.CreateClient(context.Background(), oauth2Client); err != nil {
+		t.Fatalf("Failed to create OAuth2 client: %v", err)
+	}
+
+	t.Log("Testing SSO authorization flow...")
+
+	// Step 1: Get authorization code from mock SSO
+	authCode := getMockSSOAuthCode(t, client, ssoBaseURL, testUser, testPassword)
+	t.Logf("Received auth code from SSO: %s...", authCode[:10])
+
+	// Step 2: Exchange code for token at MCP server (simulating SSO token exchange)
+	// For this test, we'll use the MCP server's own OAuth2 with user header
+	accessToken := getOAuth2TokenAuthCodeForClient(t, client, mcpBaseURL, clientID, clientSecret, testUser, signingKeyPath)
+
+	if accessToken == "" {
+		t.Fatal("Failed to get access token via SSO flow")
+	}
+
+	t.Log("Successfully obtained token via SSO flow")
+
+	// Test MCP access
+	t.Log("Testing MCP access with SSO token...")
+	testMCPInitialize(t, client, mcpBaseURL, accessToken)
+
+	t.Log("SSO integration test passed!")
+}
+
+// Helper functions
+
+func setupTestServer(t *testing.T, port int) (string, *httpserver.Server, string, string) {
```

Comment:
Switch this to use any available port, do not hardcode.  This allows multiple copies of the unit test to be run at once.